pilot/pkg/model/config.go

```diff
@@ -248,17 +248,39 @@ func resolveGatewayName(gwname string, meta config.Meta) string {
 
 // MostSpecificHostMatch compares the elements of the stack to the needle, and returns the longest stack element
 // matching the needle, or false if no element in the stack matches the needle.
-func MostSpecificHostMatch(needle host.Name, stack []host.Name) (host.Name, bool) {
+func MostSpecificHostMatch(needle host.Name, m map[host.Name]struct{}, stack []host.Name) (host.Name, bool) {
```

Comment:
I donot think in real scenario it  can have so many host names. So i doubt if it is worth this complexity

Reply:
We use ServiceEntry and WorkloadEntry to synchronize thousands of Dubbo services from Zookeeper into Mesh. The number of services was close to 4,000 in the real scenario. So we used 5,000 services in stress Testing. The time of xDS full push closed to 1m30s, which was too slow for us. And the full push blocked `debounce()`, the lock of `free` has not been released.

Reply:
Each VirtualService and DestinationRule contains three groups of blue, green, and canary, so they are complex

Reply:
Performance is improved compared to the original.

benchmark before:

```shell
➜ go test ./pilot/pkg/model/config_test.go -test.run MostSpecificHostMatch -bench=. -benchtime=10s -benchmem
goos: darwin
goarch: amd64
BenchmarkMostSpecificHostMatchExact-8            	  193176	     62688 ns/op       0 B/op	       0 allocs/op
BenchmarkMostSpecificHostMatchDestWildcard-8     	  217664	     55115 ns/op      16 B/op	       1 allocs/op
BenchmarkMostSpecificHostMatchNeedleWildcard-8   	  175539	     67566 ns/op       0 B/op	       0 allocs/op
PASS
ok  	command-line-arguments	37.928s
```

benchmark after:

```shell
➜ go test ./pilot/pkg/model/config_test.go -test.run MostSpecificHostMatch -bench=. -benchtime=10s -benchmem
goos: darwin
goarch: amd64
BenchmarkMostSpecificHostMatchExact-8            	433912726	        28.6 ns/op	       0 B/op	       0 allocs/op
BenchmarkMostSpecificHostMatchDestWildcard-8     	  309602	     38781 ns/op      16 B/op	       1 allocs/op
BenchmarkMostSpecificHostMatchNeedleWildcard-8   	  275493	     42212 ns/op       0 B/op	       0 allocs/op
PASS
ok  	command-line-arguments	39.772s
```

Reply:
For dubbo, each interface is defined as a hostname?

Reply:
Yes, we can only use ServiceEntry to plug the External Dubbo Service(interface) into Mesh. There are usually many services in the same pod. For an Dubbo service(interface), we use DestinationRule to define different subsets, such as blue, green, and canary.

Reply:
> For dubbo, each interface is defined as a hostname?

Yes , in fact, we have more than 3000 interfaces which are running in our production stage, There are too many applications (more than 1200) for these years due to rapid growth of our business , so we have to solve this scenario urgently.